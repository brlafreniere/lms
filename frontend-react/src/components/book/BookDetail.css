.book-detail-container {

}

.book-detail-image {
    background-color: gray;
    padding: 1em;
}

.book-detail-title-and-synopsis {

}