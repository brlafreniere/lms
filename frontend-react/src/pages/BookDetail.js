import React from "react";
import Axios from "axios";
import { useParams } from "react-router-dom";

export default function BookDetailPage(props) {
    let { id } = useParams();
    return (
        <BookDetail bookId={id} />
    )
}

class BookDetail extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            book: {}
        }
    }

    componentDidMount() {
        this.loadBook();
    }

    loadBook = () => {
        Axios.get(`${process.env.REACT_APP_API_URL}/books/${this.props.bookId}`).then(response => {
            this.setState({book: response.data})
        })
    }

    render() {
        return (
            <div className='book-detail'>
                <div className='book-detail-image'>
                    <div><img src={`${process.env.REACT_APP_API_URL}/uploads/${this.state.book.cover_image_file_name}`} width="200" alt={this.state.book.title + 'cover image'}/></div>
                </div>
                <div className='book-detail-title-and-synopsis'>
                    <h2>{this.state.book.title}</h2>
                    {this.state.book.synopsis}
                    <div className='book-detail-options'>
                        <button type="button" class="btn btn-primary">Check Out</button>
                    </div>
                </div>
            </div>
        )
    }
}