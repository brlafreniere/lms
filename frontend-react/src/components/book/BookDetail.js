import React from "react";
import {
    useParams,
} from "react-router-dom";

import Book from "../../modules/book";
import OverlayMenu from "../OverlayMenu"
import { BranchSelector } from "../branches/Branch";

import "./BookDetail.css"

class ReservationForm extends React.Component {
    handleSubmit = (event) => {
        event.preventDefault();
        Book.createReservation({
            book_id: this.props.book.id,
            branch_id: event.target.branch_id.value
        }).then(response => {
            console.log(response)
        }).catch(error => {
            if (error.response.data.book && error.response.data.book[0] == "has already been taken") {
                this.props.reservationFailure("You have already placed a reservation for this book.")
            }
        })
        this.props.closeOverlay()
    }

    render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <h4>Select pickup location:</h4>
                <div className="form-group">
                    <BranchSelector name="branch_id" />
                </div>
                <input type="submit" className="form-control btn btn-primary" />
            </form>
        )
    }
}

export default function BookDetail(props) {
    let { id } = useParams();
    return (
        <BookDetailComponent bookId={id} />
    )
}

class BookDetailComponent extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            book: {},
            showReservationMenu: false,
            reservable: null,
            reservationFailureMessage: ""
        }
    }

    componentDidMount() {
        this.loadBook();
    }

    loadBook = () => {
        Book.fetch(this.props.bookId).then(book => {
            this.setState({book})
            this.checkReservationStatus()
        }).catch(console.log)
    }

    showReservationMenu = () => {
        this.setState({showReservationMenu: true})
    }

    closeOverlay = () => {
        this.setState({showReservationMenu: false})
    }

    ReservationMenu = (props) => {
        if (this.state.showReservationMenu) {
            return (
                <OverlayMenu closeOverlay={this.closeOverlay}>
                    <ReservationForm book={this.state.book} closeOverlay={this.closeOverlay} reservationFailure={(message) => this.setState({reservable: false, reservationFailureMessage: message})} />
                </OverlayMenu>
            )
        } else {
            return null;
        }
    }

    CoverImage = () => {
        if (this.state.book.cover_image_file_name) {
            return (
                <div className='book-detail-image'>
                    <div>
                        <img src={`${process.env.REACT_APP_API_URL}/uploads/${this.state.book.cover_image_file_name}`}
                            width="200"
                            alt={this.state.book.title + 'cover image'}/>
                    </div>
                </div>
            )
        } else {
            return null;
        }
    }

    checkReservationStatus = () => {
        Book.reservationStatus({
            book_id: this.state.book.id
        }).then(response => {
            if (response.data.reservable) {
                this.setState({reservable: true})
            } else {
                this.setState({reservable: false})
            }
        }).catch(error => {
            this.setState({reservable: false})
        })
    }

    ReserveButton = () => {
        const reserveButton = <button className="btn btn-primary mb-3" onClick={this.showReservationMenu}>Reserve</button>
        const unavailableMessage = <div className="alert alert-danger mb-3">This book is unavailable for reservation. {this.state.reservationFailureMessage}</div>
        if (this.state.reservable) {
            return reserveButton;
        } else {
            return unavailableMessage;
        }
    }

    TitleAndSynopsis = () => {
        return (
            <div className='book-detail-title-and-synopsis p-3'>
                <h2>{this.state.book.title}</h2>
                <this.ReserveButton />
                <div>{this.state.book.synopsis}</div>
            </div>
        )
    }

    render() {
        return (
            <div className='book-detail-container d-flex'>
                <this.CoverImage />
                <this.TitleAndSynopsis />
                <this.ReservationMenu />
            </div>
        )
    }
}