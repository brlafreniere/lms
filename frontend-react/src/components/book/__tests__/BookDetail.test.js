import React from "react";
import Axios from "axios";
import renderer from "react-test-renderer";

import BookDetail from "../BookDetail";
import { MemoryRouter } from "react-router-dom";

test("<BookDetail /> snapshot test", () => {
    const response = {
        data: {
            id: 1,
            title: "Foobar",
            cover_image_file_name: "foo.jpg"
        }
    }

    Axios.get = jest.fn()
    Axios.get.mockResolvedValue(response);

    const component = renderer.create(
        <MemoryRouter>
            <BookDetail />
        </MemoryRouter>
    )

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});