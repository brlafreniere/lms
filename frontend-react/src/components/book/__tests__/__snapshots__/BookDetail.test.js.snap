// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookDetail /> snapshot test 1`] = `
<div
  className="book-detail"
>
  <div
    className="book-detail-image"
  >
    <div>
      <img
        alt="undefinedcover image"
        src="undefined/uploads/undefined"
        width="200"
      />
    </div>
  </div>
  <div
    className="book-detail-title-and-synopsis"
  >
    <h2 />
    <div
      className="book-detail-options"
    >
      <a
        className="btn btn-primary"
        href="/books/undefined/checkout"
        onClick={[Function]}
      >
        Check Out
      </a>
    </div>
  </div>
</div>
`;
